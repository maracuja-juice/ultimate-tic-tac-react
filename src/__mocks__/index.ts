import {
  unfinishedBoardMock,
  movesForUnfinishedBoardMock,
} from './unfinishedBoardMock';
import {
  crossFinishedGameMock,
  circleFinishedGameMock,
} from './finishedGameDataMock';
import {
  circleFinishedBoardMock,
  crossFinishedBoardMock,
  movesForCircleFinishedBoardMock,
  movesForCrossFinishedBoardMock,
} from './finishedBoardMock';
import {
  activeBoardsForBoardWithTwoMovesMock,
  boardWithTwoMovesMock,
} from './boardWithTwoMovesMock';
import {
  boardWithThreeMovesMock,
  activeBoardsForBoardWithThreeMovesMock,
  movesForBoardWithThreeMovesMock,
} from './boardWithThreeMovesMock';
import {
  boardWithOneMoveMock,
  activeBoardsForBoardWithOneMoveMock,
} from './boardWithOneMoveMock';

export {
  unfinishedBoardMock,
  movesForUnfinishedBoardMock,
  crossFinishedGameMock,
  circleFinishedGameMock,
  circleFinishedBoardMock,
  movesForCircleFinishedBoardMock,
  crossFinishedBoardMock,
  movesForCrossFinishedBoardMock,
  boardWithTwoMovesMock,
  activeBoardsForBoardWithTwoMovesMock,
  boardWithThreeMovesMock,
  movesForBoardWithThreeMovesMock,
  activeBoardsForBoardWithThreeMovesMock,
  boardWithOneMoveMock,
  activeBoardsForBoardWithOneMoveMock,
};
