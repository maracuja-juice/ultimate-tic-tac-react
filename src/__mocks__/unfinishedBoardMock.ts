// this not a finished board
/*
 with the following situation: ( . = not finished )
 X X .
 . O .
 . O .
 */
const unfinishedBoardMock = [
  {
    value: 0,
    position: {
      x: 0,
      y: 0,
    },
    tiles: [
      {
        boardPosition: {
          x: 0,
          y: 0,
        },
        position: {
          x: 0,
          y: 0,
        },
        value: 0,
      },
      {
        boardPosition: {
          x: 0,
          y: 0,
        },
        position: {
          x: 0,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 0,
        },
        position: {
          x: 0,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 0,
        },
        position: {
          x: 1,
          y: 0,
        },
        value: 0,
      },
      {
        boardPosition: {
          x: 0,
          y: 0,
        },
        position: {
          x: 1,
          y: 1,
        },
        value: 1,
      },
      {
        boardPosition: {
          x: 0,
          y: 0,
        },
        position: {
          x: 1,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 0,
        },
        position: {
          x: 2,
          y: 0,
        },
        value: 0,
      },
      {
        boardPosition: {
          x: 0,
          y: 0,
        },
        position: {
          x: 2,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 0,
        },
        position: {
          x: 2,
          y: 2,
        },
        value: 2,
      },
    ],
  },
  {
    value: 0,
    position: {
      x: 0,
      y: 1,
    },
    tiles: [
      {
        boardPosition: {
          x: 0,
          y: 1,
        },
        position: {
          x: 0,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 1,
        },
        position: {
          x: 0,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 1,
        },
        position: {
          x: 0,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 1,
        },
        position: {
          x: 1,
          y: 0,
        },
        value: 0,
      },
      {
        boardPosition: {
          x: 0,
          y: 1,
        },
        position: {
          x: 1,
          y: 1,
        },
        value: 0,
      },
      {
        boardPosition: {
          x: 0,
          y: 1,
        },
        position: {
          x: 1,
          y: 2,
        },
        value: 0,
      },
      {
        boardPosition: {
          x: 0,
          y: 1,
        },
        position: {
          x: 2,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 1,
        },
        position: {
          x: 2,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 1,
        },
        position: {
          x: 2,
          y: 2,
        },
        value: 2,
      },
    ],
  },
  {
    value: 2,
    position: {
      x: 0,
      y: 2,
    },
    tiles: [
      {
        boardPosition: {
          x: 0,
          y: 2,
        },
        position: {
          x: 0,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 2,
        },
        position: {
          x: 0,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 2,
        },
        position: {
          x: 0,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 2,
        },
        position: {
          x: 1,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 2,
        },
        position: {
          x: 1,
          y: 1,
        },
        value: 0,
      },
      {
        boardPosition: {
          x: 0,
          y: 2,
        },
        position: {
          x: 1,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 2,
        },
        position: {
          x: 2,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 2,
        },
        position: {
          x: 2,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 0,
          y: 2,
        },
        position: {
          x: 2,
          y: 2,
        },
        value: 2,
      },
    ],
  },
  {
    value: 2,
    position: {
      x: 1,
      y: 0,
    },
    tiles: [
      {
        boardPosition: {
          x: 1,
          y: 0,
        },
        position: {
          x: 0,
          y: 0,
        },
        value: 1,
      },
      {
        boardPosition: {
          x: 1,
          y: 0,
        },
        position: {
          x: 0,
          y: 1,
        },
        value: 1,
      },
      {
        boardPosition: {
          x: 1,
          y: 0,
        },
        position: {
          x: 0,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 0,
        },
        position: {
          x: 1,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 0,
        },
        position: {
          x: 1,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 0,
        },
        position: {
          x: 1,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 0,
        },
        position: {
          x: 2,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 0,
        },
        position: {
          x: 2,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 0,
        },
        position: {
          x: 2,
          y: 2,
        },
        value: 2,
      },
    ],
  },
  {
    value: 1,
    position: {
      x: 1,
      y: 1,
    },
    tiles: [
      {
        boardPosition: {
          x: 1,
          y: 1,
        },
        position: {
          x: 0,
          y: 0,
        },
        value: 1,
      },
      {
        boardPosition: {
          x: 1,
          y: 1,
        },
        position: {
          x: 0,
          y: 1,
        },
        value: 1,
      },
      {
        boardPosition: {
          x: 1,
          y: 1,
        },
        position: {
          x: 0,
          y: 2,
        },
        value: 1,
      },
      {
        boardPosition: {
          x: 1,
          y: 1,
        },
        position: {
          x: 1,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 1,
        },
        position: {
          x: 1,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 1,
        },
        position: {
          x: 1,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 1,
        },
        position: {
          x: 2,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 1,
        },
        position: {
          x: 2,
          y: 1,
        },
        value: 0,
      },
      {
        boardPosition: {
          x: 1,
          y: 1,
        },
        position: {
          x: 2,
          y: 2,
        },
        value: 2,
      },
    ],
  },
  {
    value: 2,
    position: {
      x: 1,
      y: 2,
    },
    tiles: [
      {
        boardPosition: {
          x: 1,
          y: 2,
        },
        position: {
          x: 0,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 2,
        },
        position: {
          x: 0,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 2,
        },
        position: {
          x: 0,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 2,
        },
        position: {
          x: 1,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 2,
        },
        position: {
          x: 1,
          y: 1,
        },
        value: 0,
      },
      {
        boardPosition: {
          x: 1,
          y: 2,
        },
        position: {
          x: 1,
          y: 2,
        },
        value: 1,
      },
      {
        boardPosition: {
          x: 1,
          y: 2,
        },
        position: {
          x: 2,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 1,
          y: 2,
        },
        position: {
          x: 2,
          y: 1,
        },
        value: 0,
      },
      {
        boardPosition: {
          x: 1,
          y: 2,
        },
        position: {
          x: 2,
          y: 2,
        },
        value: 2,
      },
    ],
  },
  {
    value: 2,
    position: {
      x: 2,
      y: 0,
    },
    tiles: [
      {
        boardPosition: {
          x: 2,
          y: 0,
        },
        position: {
          x: 0,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 0,
        },
        position: {
          x: 0,
          y: 1,
        },
        value: 1,
      },
      {
        boardPosition: {
          x: 2,
          y: 0,
        },
        position: {
          x: 0,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 0,
        },
        position: {
          x: 1,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 0,
        },
        position: {
          x: 1,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 0,
        },
        position: {
          x: 1,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 0,
        },
        position: {
          x: 2,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 0,
        },
        position: {
          x: 2,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 0,
        },
        position: {
          x: 2,
          y: 2,
        },
        value: 2,
      },
    ],
  },
  {
    value: 1,
    position: {
      x: 2,
      y: 1,
    },
    tiles: [
      {
        boardPosition: {
          x: 2,
          y: 1,
        },
        position: {
          x: 0,
          y: 0,
        },
        value: 1,
      },
      {
        boardPosition: {
          x: 2,
          y: 1,
        },
        position: {
          x: 0,
          y: 1,
        },
        value: 1,
      },
      {
        boardPosition: {
          x: 2,
          y: 1,
        },
        position: {
          x: 0,
          y: 2,
        },
        value: 1,
      },
      {
        boardPosition: {
          x: 2,
          y: 1,
        },
        position: {
          x: 1,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 1,
        },
        position: {
          x: 1,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 1,
        },
        position: {
          x: 1,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 1,
        },
        position: {
          x: 2,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 1,
        },
        position: {
          x: 2,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 1,
        },
        position: {
          x: 2,
          y: 2,
        },
        value: 2,
      },
    ],
  },
  {
    value: 2,
    position: {
      x: 2,
      y: 2,
    },
    tiles: [
      {
        boardPosition: {
          x: 2,
          y: 2,
        },
        position: {
          x: 0,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 2,
        },
        position: {
          x: 0,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 2,
        },
        position: {
          x: 0,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 2,
        },
        position: {
          x: 1,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 2,
        },
        position: {
          x: 1,
          y: 1,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 2,
        },
        position: {
          x: 1,
          y: 2,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 2,
        },
        position: {
          x: 2,
          y: 0,
        },
        value: 2,
      },
      {
        boardPosition: {
          x: 2,
          y: 2,
        },
        position: {
          x: 2,
          y: 1,
        },
        value: 0,
      },
      {
        boardPosition: {
          x: 2,
          y: 2,
        },
        position: {
          x: 2,
          y: 2,
        },
        value: 2,
      },
    ],
  },
];
export default unfinishedBoardMock;
