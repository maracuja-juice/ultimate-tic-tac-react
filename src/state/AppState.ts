import { Point } from '../util';

export interface AppState {
  currentGame: GameInformation;
  finishedGames: FinishedGameState[];
  analysisGame: AnalysisGame;
}

export interface GameInformation {
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
}

export interface FinishedGameState {
  id?: string;
  date: string;
  winner: Winner;
  gameState: SmallBoardInformation[];
  moves: MoveState[];
  saveState: string;
  errorMessage: string;
}

export interface AnalysisGame {
  id: string;
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
  currentMove: number;
}

export interface GameState {
  currentPlayer: Player;
  isFinished: boolean;
  winningPlayer: Winner;
}

export interface TileInformation {
  position: Point;
  value: TileValue;
}

export interface SmallBoardInformation extends TileInformation {
  tiles: SmallTileInformation[];
}

export interface SmallTileInformation extends TileInformation {
  boardPosition: Point;
}

export enum TileValue {
  Cross = 0,
  Circle = 1,
  Empty = 2,
  Destroyed = 3,
}

export enum Winner {
  Cross = 0,
  Circle = 1,
  None = 2,
  Draw = 3,
}

export enum Player {
  Cross = 0,
  Circle = 1,
}

export interface Move {
  boardPosition: Point;
  tilePosition: Point;
}

export interface MoveState extends Move {
  player: Player;
  moveNumber: number;
}

export interface GenericAction {
  type: string;
  // tslint:disable-next-line: no-any
  payload?: any;
}

export interface MarkSpecially {
  condition: boolean;
  position?: { boardPosition: Point; tilePosition: Point };
}
