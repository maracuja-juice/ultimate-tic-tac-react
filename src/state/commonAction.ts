export const RESTART_GAME = 'common/restart-game';

export const restartGame = () => ({
    type: RESTART_GAME
});