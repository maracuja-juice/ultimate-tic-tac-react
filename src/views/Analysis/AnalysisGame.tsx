import * as React from 'react';
import { BigBoard } from '../../components/Board/BigBoard/BigBoard';
import { RouteComponentProps } from 'react-router';
import { connect } from 'react-redux';
import {
  AppState,
  Move,
  Player,
  SmallBoardInformation,
} from '../../state/AppState';
import {
  loadFinishedGame,
  moveBackwardInHistory,
  moveForwardInHistory,
} from '../../state/analysisGame/analysisGameActions';
import './AnalysisGame.css';
import { Point } from '../../util/Point';
import { HistoryButtons } from '../../components/Analysis/HistoryButtons/HistoryButtons';
import { MoveList } from '../../components/Analysis/MoveList/MoveList';
import { scroller } from 'react-scroll/modules';
import { moveScrollElementBaseName } from '../../components/Analysis/ScrollElementConstants';

interface AnalysisGameProps extends RouteComponentProps<{ id: string }> {
  onLoad: (id: string) => void;
  moveForwardInHistory: (numberOfMoves: number) => void;
  moveBackwardInHistory: (numberOfMoves: number) => void;
  reversedMoves?: Move[];
  board: SmallBoardInformation[];
  activeBoards: Point[];
  currentPlayer: Player;
  currentMove: number;
}

export class AnalysisGame extends React.Component<AnalysisGameProps> {
  componentDidMount = () => {
    const id = this.props.match.params.id;
    this.props.onLoad(id);
  };

  scrollToElement = (moveNumberToScrollTo: number) => {
    scroller.scrollTo(moveScrollElementBaseName + moveNumberToScrollTo, {
      duration: 300,
      smooth: true,
      containerId: 'moveList',
      offset: -108,
    });
  };

  render() {
    const {
      reversedMoves,
      board,
      activeBoards,
      currentPlayer,
      currentMove,
    } = this.props;

    const currentlyAppliedMove =
      reversedMoves &&
      reversedMoves[
        reversedMoves.findIndex((m) => m.moveNumber === currentMove)!
      ];

    return (
      <div className="center">
        <div className="analysisLayout">
          <div id="moveList" className="moveList">
            {reversedMoves && (
              <MoveList
                currentMove={currentMove}
                reversedMoves={reversedMoves}
                moveForwardInHistory={this.props.moveForwardInHistory}
                moveBackwardInHistory={this.props.moveBackwardInHistory}
              />
            )}
          </div>
          <div className="historyButtons">
            {reversedMoves && (
              <HistoryButtons
                currentMove={currentMove}
                lastMove={reversedMoves[0] && reversedMoves[0].moveNumber}
                moveForwardInHistory={this.props.moveForwardInHistory}
                moveBackwardInHistory={this.props.moveBackwardInHistory}
                onInteraction={this.scrollToElement}
              />
            )}
          </div>
          <div className="analysisGame">
            {reversedMoves && (
              <BigBoard
                // tslint:disable-next-line:no-empty
                onPlayerMoved={() => {}}
                board={board}
                activeBoards={activeBoards}
                currentPlayer={currentPlayer}
                markTileSpecially={{
                  condition: currentlyAppliedMove !== undefined,
                  position: currentlyAppliedMove
                    ? {
                        boardPosition: currentlyAppliedMove.boardPosition,
                        tilePosition: currentlyAppliedMove.tilePosition,
                      }
                    : undefined,
                }}
              />
            )}
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  reversedMoves: state.analysisGame.moves.slice().reverse(),
  board: state.analysisGame.board,
  activeBoards: state.analysisGame.activeBoards,
  currentPlayer: state.analysisGame.game.currentPlayer,
  currentMove: state.analysisGame.currentMove,
});

// tslint:disable-next-line: no-any
const mapDispatchToProps = (dispatch: any) => ({
  onLoad: (id: string) => dispatch(loadFinishedGame(id)),
  moveForwardInHistory: (numberOfMoves: number) =>
    dispatch(moveForwardInHistory(numberOfMoves)),
  moveBackwardInHistory: (numberOfMoves: number) =>
    dispatch(moveBackwardInHistory(numberOfMoves)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AnalysisGame);
