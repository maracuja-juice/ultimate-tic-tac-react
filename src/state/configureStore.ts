import { AppState } from './AppState';
import { applyMiddleware, combineReducers, createStore } from 'redux';
import logger from 'redux-logger';
import tileReducer from './tiles/tileReducer';
import { composeWithDevTools } from 'redux-devtools-extension';
import gameReducer from './game/gameReducer';
import moveReducer from './moves/moveReducer';
import { enableBatching } from 'redux-batched-actions';
import activeBoardReducer from './activeBoards/activeBoardsReducer';
import createSagaMiddleware from 'redux-saga';
import { all, fork } from 'redux-saga/effects';
import moveValidationSaga from './sagas/MoveValidationSaga';

const rootreducer = combineReducers<AppState>(
    {
        game: gameReducer,
        tiles: tileReducer,
        moves: moveReducer,
        activeBoards: activeBoardReducer
    } );

export function configureStore() {

    const sagaMiddleware = createSagaMiddleware();

    const middleWaresToApply = [
        logger,
        sagaMiddleware
    ];
    const middleware = applyMiddleware( ...middleWaresToApply );
    const store = createStore( enableBatching( rootreducer ), composeWithDevTools(
        middleware
    ) );

    sagaMiddleware.run( rootSaga );

    return store;
}

function* rootSaga() {
    yield all(
        [
            fork( moveValidationSaga ),
        ]
    );
}