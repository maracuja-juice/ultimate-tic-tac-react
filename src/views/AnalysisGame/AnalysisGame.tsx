import * as React from 'react';
import { useEffect } from 'react';
import { RouteComponentProps } from 'react-router';
import { connect } from 'react-redux';
import { AnalysisGameDisplay } from './AnalysisGameDisplay';
import { parseJSON, isValid } from 'date-fns';
import { AnalysisGame, AppState } from '../../state/AppState';
import routes from '../../routes';
import {
  loadFinishedGameByDate,
  loadFinishedGameById,
  loadLatestFinishedGame,
  moveBackwardInHistory,
  moveForwardInHistory,
} from '../../state/analysisGame/analysisGameActions';
import { NoGameFound } from './NoGameFound/NoGameFound';

interface AnalysisGameRouteProps
  extends RouteComponentProps<{ param: string }> {
  analysisGame: AnalysisGame;
  loadAnalysisGameById: (id: string) => void;
  loadLatestAnalysisGame: () => void;
  loadAnalysisGameByDate: (date: Date) => void;
  moveForwardInHistory: (numberOfMoves: number) => void;
  moveBackwardInHistory: (numberOfMoves: number) => void;
}

function AnalysisGameRoute(props: AnalysisGameRouteProps) {
  // TODO move to hook
  const pathName = props.location.pathname;
  const param = props.match.params.param;
  const {
    loadLatestAnalysisGame,
    loadAnalysisGameById,
    loadAnalysisGameByDate,
    analysisGame,
  } = props;

  useEffect(() => {
    if (pathName.includes(routes.AnalysisLatest)) {
      loadLatestAnalysisGame();
    } else if (isValid(parseJSON(param))) {
      loadAnalysisGameByDate(parseJSON(param));
    } else {
      loadAnalysisGameById(param);
    }
  }, [
    param,
    pathName,
    loadAnalysisGameById,
    loadAnalysisGameByDate,
    loadLatestAnalysisGame,
  ]);

  return analysisGame.board.length > 0 ? (
    <AnalysisGameDisplay
      moveForwardInHistory={props.moveForwardInHistory}
      moveBackwardInHistory={props.moveBackwardInHistory}
      analysisGame={props.analysisGame}
    />
  ) : (
    <NoGameFound />
  );
}

const mapStateToProps = (state: AppState) => ({
  analysisGame: state.analysisGame,
});

const mapDispatchToProps = {
  loadAnalysisGameById: (id: string) => loadFinishedGameById(id),
  loadAnalysisGameByDate: (date: Date) => loadFinishedGameByDate(date),
  loadLatestAnalysisGame: () => loadLatestFinishedGame(),
  moveForwardInHistory: (numberOfMoves: number) =>
    moveForwardInHistory(numberOfMoves),
  moveBackwardInHistory: (numberOfMoves: number) =>
    moveBackwardInHistory(numberOfMoves),
};

export default connect(mapStateToProps, mapDispatchToProps)(AnalysisGameRoute);
