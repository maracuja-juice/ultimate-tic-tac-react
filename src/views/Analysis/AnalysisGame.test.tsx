import * as React from 'react';
import { configure, shallow } from 'enzyme';
import ReactSixteenAdapter from 'enzyme-adapter-react-16';
import { AnalysisGame } from './AnalysisGame';
import { circleFinishedBoardMock } from '../../__mocks__';
import { Player } from '../../state/AppState';

configure({ adapter: new ReactSixteenAdapter() });

describe('AnalysisGame', () => {
  it('should match snapshot', () => {
    // tslint:disable:no-empty
    const onLoad = jest.fn((id) => {});
    const moveForwardInHistory = jest.fn((numberOfMoves) => {});
    const moveBackwardInHistory = jest.fn((numberOfMoves) => {});
    const mock: any = jest.fn();
    const match = {
      isExact: true,
      params: {
        id: 'someId',
      },
      path: '',
      url: '',
    };

    const analysis = shallow(
      <AnalysisGame
        onLoad={onLoad}
        moveForwardInHistory={moveForwardInHistory}
        moveBackwardInHistory={moveBackwardInHistory}
        board={circleFinishedBoardMock}
        activeBoards={[]}
        currentPlayer={Player.Cross}
        currentMove={1}
        history={mock}
        location={mock}
        match={match}
      />,
    );

    expect(analysis).toMatchSnapshot();
  });

  it('should call onLoad with givenId when mounted', () => {
    // tslint:disable:no-empty
    const onLoad = jest.fn((id) => {});
    const moveForwardInHistory = jest.fn((numberOfMoves) => {});
    const moveBackwardInHistory = jest.fn((numberOfMoves) => {});
    const mock: any = jest.fn();
    const match = {
      isExact: true,
      params: {
        id: '123',
      },
      path: '',
      url: '',
    };

    shallow(
      <AnalysisGame
        onLoad={onLoad}
        moveForwardInHistory={moveForwardInHistory}
        moveBackwardInHistory={moveBackwardInHistory}
        board={circleFinishedBoardMock}
        activeBoards={[]}
        currentPlayer={Player.Cross}
        currentMove={1}
        history={mock}
        location={mock}
        match={match}
      />,
    );

    expect(onLoad).toHaveBeenCalledWith('123');
  });
});
