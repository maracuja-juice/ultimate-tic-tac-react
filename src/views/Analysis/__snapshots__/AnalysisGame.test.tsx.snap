// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnalysisGame should match snapshot 1`] = `
<div
  className="center"
>
  <div
    className="analysisLayout"
  >
    <div
      className="moveList"
      id="moveList"
    />
    <div
      className="historyButtons"
    />
    <div
      className="analysisGame"
    />
  </div>
</div>
`;
